src/resource_clients/request_queue.ts

```diff
@@ -96,6 +97,138 @@ export class RequestQueueClient extends ResourceClient {
+            }),
+        });
+
+        return cast(parseDateFields(pluckData(response.data)));
+    }
+
+    protected async _batchAddRequestsWithRetries(
+        requests: Array<Omit<RequestQueueClientRequestSchema, 'id'>>,
+        options: RequestQueueClientBatchAddRequestOptions = {},
+    ): Promise<RequestQueueClientBatchAddRequestsResult> {
+        ow(requests, ow.array.ofType(ow.object.partialShape({
+            id: ow.undefined,
+        })).minLength(1).maxLength(25));
+
+        ow(options, ow.object.partialShape({
+            forefront: ow.optional.boolean,
+            maxRetries: ow.optional.number.greaterThan(0),
+            delayBetweenRetriesMillis: ow.optional.number.greaterThan(0),
+        }));
+
+        const {
+            maxRetries = 9,
+            delayBetweenRetriesMillis = 500,
+        } = options;
+
+        // Keep track of the requests that remain to be processed (in parameter format)
+        let remainingRequests = requests;
+        // Keep track of the requests that have been processed (in api format)
+        let processedRequests: Array<ProcessedRequest> = [];
+        // The requests we have not been able to process in the last call
+        // ie. those we have not been able to process at all
+        let unprocessedRequests: Array<UnprocessedRequest> = [];
+        for (let i = 0; i < maxRetries; i++) {
+            const response = await this.httpClient.call({
+                url: this._url('requests/batch'),
+                method: 'POST',
+                data: remainingRequests,
+                params: this._params({
+                    forefront: options.forefront,
+                    clientKey: this.clientKey,
+                }),
+            });
+            const data = response.data.data as RequestQueueClientBatchAddRequestsResult;
+            processedRequests = [...processedRequests, ...data.processedRequests];
+            unprocessedRequests = data.unprocessedRequests;
+
+            // Get unique keys of all requests processed so far
+            const processedRequestsUniqueKeys = processedRequests.map(({ uniqueKey }) => uniqueKey);
+            // Requests remaining to be processed are the all that remain
+            // @ts-expect-error Undefined id - it seems to be added by ow validation
+            remainingRequests = requests.filter(({ uniqueKey }) => !processedRequestsUniqueKeys.includes(uniqueKey));
+
+            // Stop if all requests have been processed
+            if (remainingRequests.length === 0) {
+                break;
+            }
+
+            // Sleep for some time before trying again
+            await new Promise((resolve) => setTimeout(resolve, delayBetweenRetriesMillis));
+        }
+
+        // TODO: Not sure how to do this properly in ts-friendly way
+        const result = <unknown> { processedRequests, unprocessedRequests } as JsonValue;
+
+        return cast(parseDateFields(result));
+    }
+
+    async batchAddRequestsConcurrently(
```

Comment:
I think the naming here is really confusing, it's not clear why does `batchAddRequests` not do retries, but `batchAddRequestsConcurrently` does them. I'd say we should keep it simple and just have one function, with some sane default concurrency and max retry count, to not confuse users.

Reply:
Definitely, the client already does things that make life easier and the advanced function does what anyone would do manually anyway.